### DISCUSSION TITLE
vtherm does not take underlying's temperature into account

### DISCUSSION THREAD

**Discussion by flg** [jmcollin78/versatile_thermostat#694] [General]
Hello,

I'm trying to understand why vtherm does not take the underlying temperature into account.

_vtherm_sdb_ is an "over climate" over _sdb_ configured with "Use internal temperature of the underlying" enabled. On the following plot we can see that:

- Vtherm target temperature is 18,
- underlying target temperature is 18,
- while the difference between the underlying and the vtherm current temperatures is often more than 0.5 degree C.
- As a consequence heating turns on when the room temperature 18.8, at 14:30, and the device temperature is 17.9. 

![image](https://github.com/user-attachments/assets/2c09b83c-f305-4a75-885f-dc82ee3699d9)

I thought that "Use internal temperature of the underlying" was supposed to compensate for the different between external temperature sensor and device sensor. In this case, at 14:30, I was expecting the underlying target temperature to be :

- underlying target temperature = vtherm target temperature + (underlying temperature - external sensor temperature)
-17.1 = 18 + (17.9 - 18.8)

Why, with this configuration, the target temperatures of the vtherm and the underlying are the same while their respective temperatures are not?

Bonus question: what defined "vtherm_sdb heating" on the graph?

Here's a snapshot of the vtherm taken after the plot (unfortunately) to check the configuration:
`hvac_modes: off, heat
min_temp: 7
max_temp: 35
target_temp_step: 0.2
preset_modes: none, frost, eco, comfort, boost
current_temperature: 17.2
temperature: 16
hvac_action: idle
preset_mode: none
is_on: true
hvac_mode: heat
type: null
is_controlled_by_central_mode: false
last_central_mode: null
frost_temp: 8
eco_temp: 18
boost_temp: 28
comfort_temp: 22
frost_away_temp: 0
eco_away_temp: 0
boost_away_temp: 0
comfort_away_temp: 0
power_temp: null
target_temperature_step: 0.2
ext_current_temperature: 4.9
ac_mode: false
current_power: null
current_power_max: null
saved_preset_mode: eco
saved_target_temp: 18
saved_hvac_mode: off
motion_sensor_entity_id: null
motion_state: null
power_sensor_entity_id: null
max_power_sensor_entity_id: null
overpowering_state: null
presence_sensor_entity_id: null
presence_state: null
window_state: off
window_auto_state: off
window_bypass_state: false
window_sensor_entity_id: null
window_delay_sec: null
window_auto_enabled: false
window_auto_open_threshold: null
window_auto_close_threshold: null
window_auto_max_duration: null
window_action: window_turn_off
security_delay_min: 60
security_min_on_percent: 0.5
security_default_on_percent: 0.1
last_temperature_datetime: 2024-12-09T10:37:31.082322+01:00
last_ext_temperature_datetime: 2024-12-09T13:00:41.501381+01:00
security_state: false
minimal_activation_delay_sec: 10
device_power: 1
mean_cycle_power: null
total_energy: 0
last_update_datetime: 2024-12-09T13:25:02.215312+01:00
timezone: Europe/Zurich
temperature_unit: °C
is_device_active: false
ema_temp: 17.4
is_used_by_central_boiler: false
temperature_slope: 0
hvac_off_reason: null
is_over_climate: true
start_hvac_action_date: null
underlying_entities: climate.sdb
auto_fan_mode: auto_fan_high
current_auto_fan_mode: auto_fan_high
auto_activated_fan_mode: null
auto_deactivated_fan_mode: null
auto_regulation_use_device_temp: true
auto_start_stop_enable: false
auto_start_stop_level: auto_start_stop_none
auto_start_stop_dtmin: null
auto_start_stop_accumulated_error: 0
auto_start_stop_accumulated_error_threshold: null
follow_underlying_temp_change: true
friendly_name: vtherm_sdb
supported_features: 401`

**Comment by jmcollin78** [comment]
Hello, can you check this post ? I think you can be in the same issue : https://github.com/jmcollin78/versatile_thermostat/issues/690

**Comment by flg** [comment]
Hello Jean-Marc,

I don't think #690 has anything to do with my question.

Maybe my assumption is just wrong. I'm assuming that with the following conditions:

- external sensor temperature = 19
- underlying sensor temperature = 18
- vtherm target temperature = 20

The underlying target temperature should be : 19 (i.e. 20 - (19 - 18)).

Is this assumption correct?

**Reply by jmcollin78** [reply]
It is much more complicated than thatand it depends of your tuning and parameters. You can have a look at the algorithm here: https://github.com/jmcollin78/versatile_thermostat/blob/main/documentation/en/algorithms.md#the-self-regulation-algorithm-without-valve-control

**Reply by flg** [reply]
OK I think I understand. But then in the plot we can see the temperature oscillates around 19 instead of 18 (which is the vtherm target temperature). What can I do to fix that? I tried with and without "Use internal temperature of the underlying" enabled but in has no obvious impact.

**Reply by jmcollin78** [reply]
Add a self-regulation (this has been done for this exactly). Start at low level, wait for stabilization and if already not good, upgrade the level.

Use "Use internal temperature.. " only if the self-regulation is not enough. It generaly don't give good results.

To be more precise I will need curves as explained here: https://github.com/jmcollin78/versatile_thermostat/blob/main/documentation/en/additions.md#regulation-curves-with-plotly-to-fine-tune-your-thermostat

**Reply by flg** [reply]
Thanks, I will try the self-regulation.
